### DISCUSSION TITLE
global usage without request context

### DISCUSSION THREAD

**Discussion by stdOWL** [mikro-orm/mikro-orm#3370] [Q&A]
Hi,

We started a medium-sized project with NestJS and decided to use MikroORM instead of TypeORM.
We have a little structure problem that we could not solve.

When a user created, we send "create-user-services" message to a queue ( check : https://docs.nestjs.com/techniques/queues ) so a processor fetch that message and creates user external records etc. 

The processor, uses userService to create records, the idea of this structure is we will create multiple instantance of the API and these "create-user-service" and others will be distributed among different instances.

The problem is, when we called userService(we use mikro-orm repository) to create records, it gives the error _'using global EntityManager instance methods for context specific actions is disallowed. If you need to work with the global instance's identity map, use `allowGlobalContext` configuration option or `fork()` instead.'_   
When i made a research, we saw that "allowGlobalContext" is not allowed, so seems like we have to fork but problem is this userService are being used from controller too. 

So how can we solve this problem, use userService(mikro-orm calls) from controller & a global service.

**Comment by thomaschaaf** [comment]
What was the reason for mentioning me?

**Comment by stdOWL** [comment]
> What was the reason for mentioning me?

Sorry by mistaken.

I was trying to mention @B4nan

**Comment by B4nan** [comment]
Try this maybe? Or fork explicitly as the error message suggests. 

https://mikro-orm.io/docs/identity-map#userequestcontext-decorator

If thats not enough, you will need to provide some code.

**Comment by stdOWL** [comment]
Hi,

let me share the code

userService.ts : it includes all user operations like createUser etc.  when a UserEntity created in "createUser" function, it calls "createUserWallets" , and it adds "create-user-wallets" task to queue. the queue is "@nestjs/bull"

```ts

async createUser(
    userRegisterDto: UserRegisterDto
  ): Promise<UserEntity> {
    const user = this.userRepository.create(userRegisterDto);
    await this.userRepository.persistAndFlush(user);
     await this.createWallets(user);
    return user;
  }
async createUserWallets(user) {
    this.userOperationQueue.add('create-user-wallets', {
      user
    });
  }

async createSettings(
    user: UserEntity
  ): Promise<UserSettingsEntity> {
    const userSettingsEntity = new UserSettingsEntity();
    userSettingsEntity.user = user;

    await this.userSettingsRepository.persistAndFlush(userSettingsEntity);
    return userSettingsEntity;
  }

 async createWallets(
    user: UserEntity
  ): Promise<any> {


    const addresses = await this.createAddresses(user);
    const assets = await this.assetService.findAllAssets();
    const balances = Array<UserBalancesEntity>();
    for (let i = 0; i < assets.length; i++) {
      const userBalance = this.userBalancesRepository.create({
        asset: assets[i],
        user: user
      });
      this.userBalancesRepository.persist(userBalance);
      balances.push(userBalance);
    }
    await this.userBalancesRepository.flush();
    return { balances, addresses };
  }
```


the queue worker, user.worker.ts it consumes the queue.

```ts

import { UserService } from './user.service';
import { Process, Processor } from "@nestjs/bull";
import { Job } from "bull";
import { AssetService } from "../asset/asset.service";
import { UserBalancesRepository } from "./user-balances.entity";
import { UserSettingsRepository } from "./user-settings.entity";
import { UserEntity, UserRepository } from "./user.entity";
import { UseRequestContext } from '@mikro-orm/core';


@Processor('user-queue')
export class UserWorker {
  constructor(
    private userRepository: UserRepository,
    private userBalancesRepository: UserBalancesRepository,
    private userSettingsRepository: UserSettingsRepository,
    private userService: UserService,
    private assetService: AssetService,
  ) { }


  @Process('create-user-wallets')
  async createUserWallets(job: Job<unknown>) {
    const user: UserEntity = job.data.user;
    try {
      const balances = await this.userService.createBalances(user);
      const addresses = await this.userService.createAddresses(user);
      console.log(balances, addresses);
    } catch (e) {
      console.error(e);
    }

  }
}
```

when user.worker.ts tries to call userService, it gives the error. I tried to add @UseRequestContext to 'async createUserWallets(job: Job<unknown>)' in worker but did not work. ( queue stop consuming the tasks )


i tested this code;
```ts

@Process('create-user-wallets')
  @UseRequestContext()
  async createUserWallets(job: Job<unknown>) {
    const user: UserEntity = job.data.user;
    try {
      const addresses = await this.userService.createAddresses(user);
      console.log(addresses);
    } catch (e) {
      console.error(e);
    }

  }
```

in userService 'createAddresses' function, first finds all networks by calling networkService the code is : 

```ts

async createAddresses(
    user: UserEntity
  ): Promise<UserBalancesEntity[]> {
    const addresses = Array<UserBalancesEntity>();

    const networks = await this.networkService.findAllNetworks();
    for (let i = 0; i < networks.length; i++) {
      // TODO: fill address info
      const address = null;
      const secret = null;

      const userAddress = this.userAddressesRepository.create({
        network: networks[i],
        user: user,
        address,
        secret
      });
      this.userAddressesRepository.persist(userAddress);
      addresses.push(userAddress);
    }
    await this.userAddressesRepository.flush();
    return addresses;



  }
```

when mikro-orm tried to call this sql query; 'select "n0".* from "networks" as "n0" '  it gives 

'TypeError: Cannot read property 'getPrimaryKey' of undefined' it happens because NetworkEntity has OneToOne relation with AssetEntity
```ts
@OneToOne(() => AssetEntity, (asset) => asset.nativeNetwork, {
    owner: true,
    mapToPk: true,
    onDelete: 'cascade',
    onUpdateIntegrity: 'cascade',
    nullable: true,
  })
  nativeAsset?: AssetEntity;
```

and, i guess, it does not fetch it when i call service with @UseRequestContext decorator.

the error dump about getPrimaryKey error;

if (data.nativeAsset != null && convertCustomTypes) {
      data.nativeAsset = convertToDatabaseValue_nativeAsset(entity.nativeAsset.__helper.getPrimaryKey());
>   }

**Reply by B4nan** [reply]
> @OneToOne(() => AssetEntity, (asset) => asset.nativeNetwork, {
>    owner: true,
>    mapToPk: true,
>    onDelete: 'cascade',
>    onUpdateIntegrity: 'cascade',
>    nullable: true,
>  })
>  nativeAsset?: AssetEntity;

This is wrong, you can't map to `AssetEntity` if you use `mapToPk`, as that literally means you are mapping to its PK, not to the entity type.

> @Process('create-user-wallets')
>  @UseRequestContext()

This might be wrong as well, the order of decorators matter (but maybe you picked the right one, idk).

https://mikro-orm.io/docs/usage-with-nestjs#request-scoped-handlers-in-queues
